core/src/main/scala/cats/Eval.scala

```diff
@@ -134,23 +136,57 @@ final case class Now[A](value: A) extends Eval[A] {
  * by the closure) will not be retained, and will be available for
  * garbage collection.
  */
-final class Later[A](f: () => A) extends Eval[A] {
-  private[this] var thunk: () => A = f
-
-  // The idea here is that `f` may have captured very large
-  // structures, but produce a very small result. In this case, once
-  // we've calculated a value, we would prefer to be able to free
-  // everything else.
-  //
-  // (For situations where `f` is small, but the output will be very
-  // expensive to store, consider using `Always`.)
-  lazy val value: A = {
-    val result = thunk()
-    thunk = null // scalastyle:off
-    result
-  }
+final class Later[A](f: () => A) extends Eval[A] with Externalizable {
+
+  // we use lock to prevent possible problems with someone else
+  // synchronizing on 'this'.
+  private[this] val lock = new Object()
+  private[this] var thunk: Function0[A] = f
+  private[this] var result: A = _
+
+  final def value: A =
+    if (thunk == null) {
+      result
+    } else lock.synchronized {
+      if (thunk == null) result else {
```

Comment:
I think that double-checked locking on the JVM requires a `volatile` var because of...reasons. http://stackoverflow.com/questions/7855700/why-is-volatile-used-in-this-example-of-double-checked-locking hits on this. You are definitely more of a JVM expert than I am though, so I feel like you could probably convince me that this isn't a concern.